paddle/phi/kernels/onednn/matmul_kernel.cc

```diff
@@ -0,0 +1,450 @@
+                     {DNNL_ARG_DST, *output_}});
+    astream.wait();
+  }
+
+  template <typename T>
+  DenseTensor UpdateDataFormat(const DenseTensor *data,
+                               int num_col_dims,
+                               const OneDNNContext &ctx) {
+    DenseTensor x_tmp;
+    DenseTensor data_matrix;
+    // This code is enforcing plain (non-blocked) memory arrangement
+    // in order to flatten (reduce dimensionality) of DenseTensor later
+    auto src_mdesc = data->mem_desc();
+    auto dst_mdesc = data->dims().size() >= 4
+                         ? (data->dims().size() == 5
+                                ? CreateMemDescriptor<T>(
+                                      data, funcs::OneDNNMemoryFormat::ncdhw)
+                                : CreateMemDescriptor<T>(
+                                      data, funcs::OneDNNMemoryFormat::nchw))
+                         : src_mdesc;
+
+    if (src_mdesc != dst_mdesc) {
+      x_tmp.mutable_data<T>(ctx.GetPlace(), data->memory_size());
+
+      Reorder(src_mdesc,
+              dst_mdesc,
+              phi::funcs::to_void_cast<T>(data->data<T>()),
+              phi::funcs::to_void_cast<T>(x_tmp.data<T>()));
+
+      x_tmp.Resize(data->dims());
+      x_tmp.set_mem_desc(dst_mdesc);
+      data_matrix = paddle::framework::ReshapeToMatrix(x_tmp, num_col_dims);
+    } else {
+      data_matrix = paddle::framework::ReshapeToMatrix(*data, num_col_dims);
+    }
+
+    return data_matrix;
+  }
+
+  void UpdateDataPointers(const OneDNNContext &ctx,
+                          DenseTensor *out,
+                          const DenseTensor *in) {
+    x_input_->set_data_handle(phi::funcs::to_void_cast<XT>(in->data<XT>()));
+    output_->set_data_handle(out->mutable_data<OT>(ctx.GetPlace()));
+    out->set_mem_desc(output_->get_desc());
+  }
+
+  template <typename T>
+  memory::desc CreateMemDescriptor(
+      const DenseTensor *tensor,
+      funcs::OneDNNMemoryFormat format,
+      memory::data_type type = phi::funcs::OneDNNGetDataType<T>()) {
+    auto dims = phi::vectorize<int64_t>(tensor->dims());
+    return phi::funcs::OneDNNMemDesc(dims, type, format);
+  }
+
+  template <typename T>
+  memory::desc CreateMemDescriptor(
+      const std::vector<int64_t> &dims,
+      funcs::OneDNNMemoryFormat format,
+      memory::data_type type = phi::funcs::OneDNNGetDataType<T>()) {
+    return phi::funcs::OneDNNMemDesc(dims, type, format);
+  }
+
+  template <typename T>
+  memory CreateMemory(const memory::desc &desc, const DenseTensor *tensor) {
+    return memory(
+        desc, engine_, phi::funcs::to_void_cast<T>(tensor->data<T>()));
+  }
+
+  memory CreateDstMemory(
+      const inner_product_forward::primitive_desc &mul_prim_desc,
+      const OneDNNContext &ctx,
+      DenseTensor *output) {
+    auto dst_desc = mul_prim_desc.dst_desc();
+    auto buffer_size = dst_desc.get_size();
+
+    OT *output_data = output->mutable_data<OT>(ctx.GetPlace(), buffer_size);
```

Comment:
```suggestion
    OT *output_data = ctx.template Alloc<OT>(out, buffer_size);
```
`mutable_data` can't be used in PHI